File: frontend/src/types/trading.ts
// ── Enum 타입 ──

export type TradingMode = "paper" | "live";
export type StrategyType = "ma_crossover" | "mean_reversion" | "custom";
export type OrderSide = "buy" | "sell";
export type OrderType = "market" | "limit";
export type OrderStatus = "pending" | "submitted" | "filled" | "partial" | "cancelled" | "rejected";
export type ScheduleLogStatus = "success" | "skipped" | "error";

// ── API 응답 타입 (snake_case, 백엔드 매칭) ──

export interface TradingAccountApi {
  id: string;
  mode: TradingMode;
  initial_capital: number;
  cash_balance: number | null;
  is_active: boolean;
  allow_netting: boolean;
  token_expires_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface AutoSelectConfigApi {
  enabled: boolean;
  top_n: number;
  market: "KOSPI" | "KOSDAQ" | "ALL";
  min_volume_value: number;
  blacklist: string[];
}

export interface AutoSelectedTickersInfoApi {
  tickers: string[];
  generated_at: string;
  rule_version: string;
  details?: Array<Record<string, unknown>> | null;
}

export interface AutoTickerSelectionHistoryApi {
  id: string;
  generated_at: string;
  rule_version: string;
  triggered_by: string;
  selected_tickers: Array<Record<string, unknown>>;
  excluded_sample: Array<Record<string, unknown>> | null;
  config_snapshot: Record<string, unknown>;
}

export interface AutoTickerPreviewResponseApi {
  rule_version: string;
  selected: Array<Record<string, unknown>>;
  excluded_sample: Array<Record<string, unknown>>;
  config_snapshot: Record<string, unknown>;
  available_cash: number;
  total_eval: number;
  max_position_pct: number;
}

export interface TradingStrategyApi {
  id: string;
  account_id: string;
  name: string;
  strategy_type: StrategyType;
  params_json: Record<string, number>;
  target_tickers: string[];
  interval_minutes: number;
  is_scheduled: boolean;
  market_hours_only: boolean;
  is_active: boolean;
  initial_capital: number;
  realized_pnl: number;
  priority: number;
  auto_select_config: AutoSelectConfigApi;
  auto_selected_tickers: AutoSelectedTickersInfoApi | null;
  created_at: string;
  updated_at: string;
}

export interface TradingOrderApi {
  id: string;
  account_id: string;
  strategy_id: string | null;
  side: OrderSide;
  ticker: string;
  ticker_name: string;
  quantity: number;
  price: number;
  filled_quantity: number | null;
  filled_price: number | null;
  order_type: OrderType;
  status: OrderStatus;
  kis_order_id: string | null;
  reason: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface TradingPositionApi {
  id: string;
  account_id: string;
  ticker: string;
  ticker_name: string;
  quantity: number;
  avg_buy_price: number;
  current_price: number | null;
  unrealized_pnl: number | null;
  created_at: string;
  updated_at: string;
}

export interface TradingPerformanceApi {
  total_trades: number;
  winning_trades: number;
  losing_trades: number;
  win_rate: number;
  total_realized_pnl: number;
  total_unrealized_pnl: number;
  initial_capital: number;
  current_value: number;
  return_rate: number;
}

export interface ScheduleLogSummaryApi {
  id: string;
  status: ScheduleLogStatus;
  executed_at: string;
  completed_at: string | null;
  tickers_evaluated: number;
  orders_placed: number;
  orders_filled: number;
  skip_reason: string | null;
  error_message: string | null;
}

export interface ScheduleStatusApi {
  is_active: boolean;
  strategy_id: string | null;
  interval_minutes: number | null;
  next_run_at: string | null;
  last_run: ScheduleLogSummaryApi | null;
}

// ── 프론트엔드 타입 (camelCase) ──

export interface TradingAccount {
  id: string;
  mode: TradingMode;
  initialCapital: number;
  cashBalance: number | null;
  isActive: boolean;
  allowNetting: boolean;
  tokenExpiresAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface AutoSelectConfig {
  enabled: boolean;
  topN: number;
  market: "KOSPI" | "KOSDAQ" | "ALL";
  minVolumeValue: number;
  blacklist: string[];
}

export interface AutoSelectedTickersInfo {
  tickers: string[];
  generatedAt: string;
  ruleVersion: string;
  details?: Array<Record<string, unknown>> | null;
}

export interface AutoTickerSelectionHistory {
  id: string;
  generatedAt: string;
  ruleVersion: string;
  triggeredBy: string;
  selectedTickers: Array<Record<string, unknown>>;
  excludedSample: Array<Record<string, unknown>> | null;
  configSnapshot: Record<string, unknown>;
}

export interface AutoTickerPreviewResponse {
  ruleVersion: string;
  selected: Array<Record<string, unknown>>;
  excludedSample: Array<Record<string, unknown>>;
  configSnapshot: Record<string, unknown>;
  availableCash: number;
  totalEval: number;
  maxPositionPct: number;
}

export interface TradingStrategy {
  id: string;
  accountId: string;
  name: string;
  strategyType: StrategyType;
  paramsJson: Record<string, number>;
  targetTickers: string[];
  intervalMinutes: number;
  isScheduled: boolean;
  marketHoursOnly: boolean;
  isActive: boolean;
  initialCapital: number;
  realizedPnl: number;
  priority: number;
  autoSelectConfig: AutoSelectConfig;
  autoSelectedTickers: AutoSelectedTickersInfo | null;
  createdAt: string;
  updatedAt: string;
}

export interface TradingOrder {
  id: string;
  accountId: string;
  strategyId: string | null;
  side: OrderSide;
  ticker: string;
  tickerName: string;
  quantity: number;
  price: number;
  filledQuantity: number | null;
  filledPrice: number | null;
  orderType: OrderType;
  status: OrderStatus;
  kisOrderId: string | null;
  reason: string | null;
  errorMessage: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface TradingPosition {
  id: string;
  accountId: string;
  ticker: string;
  tickerName: string;
  quantity: number;
  avgBuyPrice: number;
  currentPrice: number | null;
  unrealizedPnl: number | null;
  createdAt: string;
  updatedAt: string;
}

export interface TradingPerformance {
  totalTrades: number;
  winningTrades: number;
  losingTrades: number;
  winRate: number;
  totalRealizedPnl: number;
  totalUnrealizedPnl: number;
  initialCapital: number;
  currentValue: number;
  returnRate: number;
}

export interface ScheduleStatus {
  isActive: boolean;
  strategyId: string | null;
  intervalMinutes: number | null;
  nextRunAt: string | null;
  lastRun: ScheduleLogSummaryApi | null;
}

// ── KIS 실시간 잔고 타입 ──

export interface KisHolding {
  ticker: string;
  name: string;
  quantity: number;
  avgPrice: number;
  currentPrice: number;
  evalAmount: number;
  pnl: number;
  pnlRate: number;
}

export interface KisBalance {
  cash: number;
  totalEval: number;
  totalPnl: number;
  holdings: KisHolding[];
}

export interface KisBalanceApi {
  cash: number;
  total_eval: number;
  total_pnl: number;
  holdings: {
    ticker: string;
    name: string;
    quantity: number;
    avg_price: number;
    current_price: number;
    eval_amount: number;
    pnl: number;
    pnl_rate: number;
  }[];
}

export function toKisBalance(api: KisBalanceApi): KisBalance {
  return {
    cash: api.cash,
    totalEval: api.total_eval,
    totalPnl: api.total_pnl,
    holdings: api.holdings.map((h) => ({
      ticker: h.ticker,
      name: h.name,
      quantity: h.quantity,
      avgPrice: h.avg_price,
      currentPrice: h.current_price,
      evalAmount: h.eval_amount,
      pnl: h.pnl,
      pnlRate: h.pnl_rate,
    })),
  };
}

// ── 요청 타입 (snake_case, API 직접 전송) ──

export interface TradingAccountCreateRequest {
  mode: TradingMode;
}

export interface TradingAccountUpdateRequest {
  allow_netting?: boolean;
}

export interface AutoSelectConfigRequest {
  enabled: boolean;
  top_n: number;
  market: "KOSPI" | "KOSDAQ" | "ALL";
  min_volume_value: number;
  blacklist: string[];
}

export interface TradingStrategyCreateRequest {
  account_id: string;
  name: string;
  strategy_type: StrategyType;
  params_json: Record<string, number>;
  target_tickers?: string[];
  interval_minutes: number;
  market_hours_only: boolean;
  priority?: number;
  initial_capital: number;
  auto_select_config?: AutoSelectConfigRequest;
}

export interface TradingStrategyUpdateRequest {
  name?: string;
  params_json?: Record<string, number>;
  target_tickers?: string[];
  interval_minutes?: number;
  market_hours_only?: boolean;
  priority?: number;
  initial_capital?: number;
  auto_select_config?: AutoSelectConfigRequest;
}

export interface RebalanceAllocationItem {
  strategy_id: string;
  initial_capital: number;
}

export interface AccountRebalanceRequest {
  allocations: RebalanceAllocationItem[];
}

export interface AccountCapitalSummary {
  accountTotal: number;
  allocated: number;
  available: number;
}

export interface AccountCapitalSummaryApi {
  account_total: number;
  allocated: number;
  available: number;
}

export function toAccountCapitalSummary(
  api: AccountCapitalSummaryApi,
): AccountCapitalSummary {
  return {
    accountTotal: Number(api.account_total),
    allocated: Number(api.allocated),
    available: Number(api.available),
  };
}

export type DepositAllocationMode = "manual" | "pro_rata" | "reserve";

export interface DepositAllocationItem {
  strategy_id: string;
  amount: number;
}

export interface AccountDepositRequest {
  amount: number;
  mode: DepositAllocationMode;
  allocations?: DepositAllocationItem[];
}

// ── 변환 함수 ──

export function toTradingAccount(api: TradingAccountApi): TradingAccount {
  return {
    id: api.id,
    mode: api.mode,
    initialCapital: api.initial_capital,
    cashBalance: api.cash_balance,
    isActive: api.is_active,
    allowNetting: api.allow_netting ?? false,
    tokenExpiresAt: api.token_expires_at,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

export function toAutoSelectConfig(api: AutoSelectConfigApi | null | undefined): AutoSelectConfig {
  return {
    enabled: api?.enabled ?? false,
    topN: api?.top_n ?? 10,
    market: api?.market ?? "ALL",
    minVolumeValue: api?.min_volume_value ?? 10_000_000_000,
    blacklist: api?.blacklist ?? [],
  };
}

export function toAutoSelectedTickersInfo(
  api: AutoSelectedTickersInfoApi | null | undefined,
): AutoSelectedTickersInfo | null {
  if (!api) return null;
  return {
    tickers: api.tickers,
    generatedAt: api.generated_at,
    ruleVersion: api.rule_version,
    details: api.details ?? null,
  };
}

export function toAutoTickerHistory(api: AutoTickerSelectionHistoryApi): AutoTickerSelectionHistory {
  return {
    id: api.id,
    generatedAt: api.generated_at,
    ruleVersion: api.rule_version,
    triggeredBy: api.triggered_by,
    selectedTickers: api.selected_tickers ?? [],
    excludedSample: api.excluded_sample ?? null,
    configSnapshot: api.config_snapshot ?? {},
  };
}

export function toAutoTickerPreview(api: AutoTickerPreviewResponseApi): AutoTickerPreviewResponse {
  return {
    ruleVersion: api.rule_version,
    selected: api.selected ?? [],
    excludedSample: api.excluded_sample ?? [],
    configSnapshot: api.config_snapshot ?? {},
    availableCash: api.available_cash,
    totalEval: api.total_eval,
    maxPositionPct: api.max_position_pct,
  };
}

export function toTradingStrategy(api: TradingStrategyApi): TradingStrategy {
  return {
    id: api.id,
    accountId: api.account_id,
    name: api.name,
    strategyType: api.strategy_type,
    paramsJson: api.params_json,
    targetTickers: api.target_tickers,
    intervalMinutes: api.interval_minutes,
    isScheduled: api.is_scheduled,
    marketHoursOnly: api.market_hours_only,
    isActive: api.is_active,
    initialCapital: api.initial_capital ?? 0,
    realizedPnl: api.realized_pnl ?? 0,
    priority: api.priority ?? 0,
    autoSelectConfig: toAutoSelectConfig(api.auto_select_config),
    autoSelectedTickers: toAutoSelectedTickersInfo(api.auto_selected_tickers),
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

export function toTradingOrder(api: TradingOrderApi): TradingOrder {
  return {
    id: api.id,
    accountId: api.account_id,
    strategyId: api.strategy_id,
    side: api.side,
    ticker: api.ticker,
    tickerName: api.ticker_name,
    quantity: api.quantity,
    price: api.price,
    filledQuantity: api.filled_quantity,
    filledPrice: api.filled_price,
    orderType: api.order_type,
    status: api.status,
    kisOrderId: api.kis_order_id,
    reason: api.reason,
    errorMessage: api.error_message,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

export function toTradingPosition(api: TradingPositionApi): TradingPosition {
  return {
    id: api.id,
    accountId: api.account_id,
    ticker: api.ticker,
    tickerName: api.ticker_name,
    quantity: api.quantity,
    avgBuyPrice: api.avg_buy_price,
    currentPrice: api.current_price,
    unrealizedPnl: api.unrealized_pnl,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

export function toTradingPerformance(api: TradingPerformanceApi): TradingPerformance {
  return {
    totalTrades: api.total_trades,
    winningTrades: api.winning_trades,
    losingTrades: api.losing_trades,
    winRate: api.win_rate,
    totalRealizedPnl: api.total_realized_pnl,
    totalUnrealizedPnl: api.total_unrealized_pnl,
    initialCapital: api.initial_capital,
    currentValue: api.current_value,
    returnRate: api.return_rate,
  };
}

export function toScheduleStatus(api: ScheduleStatusApi): ScheduleStatus {
  return {
    isActive: api.is_active,
    strategyId: api.strategy_id,
    intervalMinutes: api.interval_minutes,
    nextRunAt: api.next_run_at,
    lastRun: api.last_run,
  };
}

// ── 라벨맵 ──

export const tradingModeLabels: Record<TradingMode, string> = {
  paper: "모의투자",
  live: "실전",
};

export const strategyTypeLabels: Record<StrategyType, string> = {
  ma_crossover: "이동평균 교차",
  mean_reversion: "평균회귀",
  custom: "커스텀",
};

export const orderSideLabels: Record<OrderSide, string> = {
  buy: "매수",
  sell: "매도",
};

export const orderStatusLabels: Record<OrderStatus, string> = {
  pending: "대기",
  submitted: "접수",
  filled: "체결",
  partial: "부분체결",
  cancelled: "취소",
  rejected: "거부",
};

File: backend/app/schemas/trading.py
"""자동매매 스키마 — Pydantic v2"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import TYPE_CHECKING, Annotated
from uuid import UUID

from pydantic import BaseModel, Field, StringConstraints, model_validator

from app.models.trading import (
    OrderSide,
    OrderStatus,
    OrderType,
    ScheduleLogStatus,
    StrategyType,
    TradingMode,
)
from app.services.crypto_service import (
    decrypt_decimal,
    decrypt_decimal_optional,
)

if TYPE_CHECKING:
    from app.models.trading import (
        AdaptiveRule as AdaptiveRuleModel,
        TradingAccount as TradingAccountModel,
        TradingDecision as TradingDecisionModel,
        TradingOrder as TradingOrderModel,
        TradingPosition as TradingPositionModel,
        TradingScheduleLog as TradingScheduleLogModel,
        TradingStrategy as TradingStrategyModel,
    )


# ──────────────────────────────────────────────
# TradingAccount
# ──────────────────────────────────────────────

class TradingAccountCreate(BaseModel):
    mode: TradingMode = TradingMode.PAPER


class TradingAccountResponse(BaseModel):
    id: UUID
    mode: TradingMode
    initial_capital: Decimal
    cash_balance: Decimal | None
    is_active: bool
    allow_netting: bool
    token_expires_at: datetime | None
    created_at: datetime
    updated_at: datetime


class TradingAccountUpdate(BaseModel):
    allow_netting: bool | None = None


class RebalanceAllocationItem(BaseModel):
    """리밸런싱: 전략별 신규 initial_capital."""
    strategy_id: UUID
    initial_capital: Decimal = Field(ge=Decimal("0"))


class AccountRebalanceRequest(BaseModel):
    """계좌 내 활성 전략들의 initial_capital 일괄 재배분."""
    allocations: list[RebalanceAllocationItem]


class DepositAllocationMode(str, Enum):
    """입금 할당 방식."""
    MANUAL = "manual"       # 사용자가 전략별 금액을 직접 지정
    PRO_RATA = "pro_rata"   # 기존 전략 initial_capital 비율대로 자동 분배
    RESERVE = "reserve"     # 계좌 총액만 증액, 전략에는 배분하지 않음


class DepositAllocationItem(BaseModel):
    """수동 입금 분배: 전략별 추가 금액."""
    strategy_id: UUID
    amount: Decimal = Field(ge=Decimal("0"))


class AccountDepositRequest(BaseModel):
    """계좌 신규 입금 반영 + 전략 할당."""
    amount: Decimal = Field(gt=Decimal("0"))
    mode: DepositAllocationMode
    # mode=manual일 때만 사용
    allocations: list[DepositAllocationItem] | None = None

    @model_validator(mode="after")
    def _check_mode_consistency(self) -> "AccountDepositRequest":
        if self.mode == DepositAllocationMode.MANUAL:
            # MANUAL은 반드시 비어있지 않은 allocations를 요구
            if self.allocations is None or len(self.allocations) == 0:
                raise ValueError(
                    "manual 모드에서는 allocations가 비어있을 수 없습니다.",
                )
        elif self.mode in (DepositAllocationMode.PRO_RATA, DepositAllocationMode.RESERVE):
            # PRO_RATA/RESERVE는 allocations 필드 자체를 지정해서는 안 됨
            # (빈 리스트 []도 부적절한 입력으로 취급)
            if self.allocations is not None:
                raise ValueError(
                    f"{self.mode.value} 모드에서는 allocations를 지정할 수 없습니다.",
                )
        return self


def account_to_response(account: TradingAccountModel) -> TradingAccountResponse:
    return TradingAccountResponse(
        id=account.id,
        mode=account.mode,
        initial_capital=decrypt_decimal(account.initial_capital),
        cash_balance=decrypt_decimal_optional(account.cash_balance),
        is_active=account.is_active,
        allow_netting=account.allow_netting,
        token_expires_at=account.token_expires_at,
        created_at=account.created_at,
        updated_at=account.updated_at,
    )


# ──────────────────────────────────────────────
# TradingStrategy
# ──────────────────────────────────────────────

_DEFAULT_MA_PARAMS = {
    "fast_period": 5,
    "slow_period": 20,
    "rsi_period": 14,
    "rsi_overbought": 70,
    "rsi_oversold": 30,
}


TickerCode = Annotated[str, StringConstraints(pattern=r"^[0-9]{6}$")]


class AutoSelectConfig(BaseModel):
    """자동 종목 선정 설정 — 전략에 내장."""
    enabled: bool = False
    top_n: int = Field(default=10, ge=3, le=30)
    market: str = Field(default="ALL", pattern="^(KOSPI|KOSDAQ|ALL)$")
    min_volume_value: int = Field(default=10_000_000_000, ge=0)
    blacklist: list[TickerCode] = Field(default_factory=list, max_length=50)


class AutoSelectedTickersInfo(BaseModel):
    """전략에 저장된 마지막 자동 선정 결과."""
    tickers: list[str]
    generated_at: datetime
    rule_version: str
    details: list[dict] | None = None


class AutoTickerSelectionHistory(BaseModel):
    id: UUID
    generated_at: datetime
    rule_version: str
    triggered_by: str
    selected_tickers: list[dict]
    excluded_sample: list[dict] | None = None
    config_snapshot: dict


class AutoTickerPreviewResponse(BaseModel):
    """미저장 미리보기 — 현재 설정대로 돌렸을 때의 결과."""
    rule_version: str
    selected: list[dict]
    excluded_sample: list[dict]
    config_snapshot: dict
    available_cash: Decimal
    total_eval: Decimal
    max_position_pct: Decimal


class TradingStrategyCreate(BaseModel):
    account_id: UUID
    name: str = Field(max_length=100)
    strategy_type: StrategyType = StrategyType.MA_CROSSOVER
    params_json: dict = Field(default_factory=lambda: _DEFAULT_MA_PARAMS.copy())
    target_tickers: list[str] = Field(default_factory=list)
    interval_minutes: int = Field(default=10, ge=1, le=60)
    market_hours_only: bool = True
    # Phase 2: 전략별 초기 할당 자본 (필수). 0이면 매수 불가.
    initial_capital: Decimal = Field(default=Decimal("0"), ge=Decimal("0"))
    # Phase 4: 처리 우선순위 (높을수록 우선). 0~100.
    priority: int = Field(default=0, ge=0, le=100)
    auto_select_config: AutoSelectConfig = Field(default_factory=AutoSelectConfig)


class TradingStrategyUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    params_json: dict | None = None
    target_tickers: list[str] | None = None
    interval_minutes: int | None = Field(default=None, ge=1, le=60)
    market_hours_only: bool | None = None
    initial_capital: Decimal | None = Field(default=None, ge=Decimal("0"))
    priority: int | None = Field(default=None, ge=0, le=100)
    auto_select_config: AutoSelectConfig | None = None


class TradingStrategyResponse(BaseModel):
    id: UUID
    account_id: UUID
    name: str
    strategy_type: StrategyType
    params_json: dict
    target_tickers: list[str]
    interval_minutes: int
    is_scheduled: bool
    market_hours_only: bool
    is_active: bool
    initial_capital: Decimal
    realized_pnl: Decimal
    priority: int
    auto_select_config: AutoSelectConfig
    auto_selected_tickers: AutoSelectedTickersInfo | None = None
    created_at: datetime
    updated_at: datetime


def strategy_to_response(strategy: TradingStrategyModel) -> TradingStrategyResponse:
    from app.services.strategy_capital import get_initial_capital, get_realized_pnl
    cfg_raw = strategy.auto_select_config or {}
    info: AutoSelectedTickersInfo | None = None
    stored = strategy.auto_selected_tickers or None
    if stored and stored.get("generated_at"):
        try:
            info = AutoSelectedTickersInfo.model_validate(stored)
        except Exception:  # noqa: BLE001 — 스키마 불일치 레코드 방어
            info = None
    return TradingStrategyResponse(
        id=strategy.id,
        account_id=strategy.account_id,
        name=strategy.name,
        strategy_type=strategy.strategy_type,
        params_json=strategy.params_json,
        target_tickers=strategy.target_tickers,
        interval_minutes=strategy.interval_minutes,
        is_scheduled=strategy.is_scheduled,
        market_hours_only=strategy.market_hours_only,
        is_active=strategy.is_active,
        initial_capital=get_initial_capital(strategy),
        realized_pnl=get_realized_pnl(strategy),
        priority=strategy.priority,
        auto_select_config=AutoSelectConfig.model_validate(
            {
                "enabled": cfg_raw.get("enabled", False),
                "top_n": cfg_raw.get("top_n", 10),
                "market": cfg_raw.get("market", "ALL"),
                "min_volume_value": cfg_raw.get("min_volume_value", 10_000_000_000),
                "blacklist": cfg_raw.get("blacklist", []),
            }
        ),
        auto_selected_tickers=info,
        created_at=strategy.created_at,
        updated_at=strategy.updated_at,
    )


# ──────────────────────────────────────────────
# TradingOrder
# ──────────────────────────────────────────────

class TradingOrderResponse(BaseModel):
    id: UUID
    account_id: UUID
    strategy_id: UUID | None
    side: OrderSide
    ticker: str
    ticker_name: str
    quantity: Decimal
    price: Decimal
    filled_quantity: Decimal | None
    filled_price: Decimal | None
    order_type: OrderType
    status: OrderStatus
    kis_order_id: str | None
    reason: str | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime


def order_to_response(order: TradingOrderModel) -> TradingOrderResponse:
    return TradingOrderResponse(
        id=order.id,
        account_id=order.account_id,
        strategy_id=order.strategy_id,
        side=order.side,
        ticker=order.ticker,
        ticker_name=order.ticker_name,
        quantity=decrypt_decimal(order.quantity),
        price=decrypt_decimal(order.price),
        filled_quantity=decrypt_decimal_optional(order.filled_quantity),
        filled_price=decrypt_decimal_optional(order.filled_price),
        order_type=order.order_type,
        status=order.status,
        kis_order_id=order.kis_order_id,
        reason=order.reason,
        error_message=order.error_message,
        created_at=order.created_at,
        updated_at=order.updated_at,
    )


# ──────────────────────────────────────────────
# TradingPosition
# ──────────────────────────────────────────────

class TradingPositionResponse(BaseModel):
    id: UUID
    account_id: UUID
    strategy_id: UUID
    ticker: str
    ticker_name: str
    quantity: Decimal
    avg_buy_price: Decimal
    current_price: Decimal | None
    unrealized_pnl: Decimal | None
    created_at: datetime
    updated_at: datetime


def position_to_response(position: TradingPositionModel) -> TradingPositionResponse:
    return TradingPositionResponse(
        id=position.id,
        account_id=position.account_id,
        strategy_id=position.strategy_id,
        ticker=position.ticker,
        ticker_name=position.ticker_name,
        quantity=decrypt_decimal(position.quantity),
        avg_buy_price=decrypt_decimal(position.avg_buy_price),
        current_price=Decimal(str(position.current_price)) if position.current_price else None,
        unrealized_pnl=decrypt_decimal_optional(position.unrealized_pnl),
        created_at=position.created_at,
        updated_at=position.updated_at,
    )


# ──────────────────────────────────────────────
# Schedule
# ──────────────────────────────────────────────

class ScheduleStartRequest(BaseModel):
    strategy_id: UUID


class ScheduleStatusResponse(BaseModel):
    is_active: bool
    strategy_id: UUID | None = None
    interval_minutes: int | None = None
    next_run_at: datetime | None = None
    last_run: ScheduleLogSummary | None = None


class ScheduleLogSummary(BaseModel):
    id: UUID
    status: ScheduleLogStatus
    executed_at: datetime
    completed_at: datetime | None
    tickers_evaluated: int
    orders_placed: int
    orders_filled: int
    skip_reason: str | None
    error_message: str | None


def schedule_log_to_summary(log: TradingScheduleLogModel) -> ScheduleLogSummary:
    return ScheduleLogSummary(
        id=log.id,
        status=log.status,
        executed_at=log.executed_at,
        completed_at=log.completed_at,
        tickers_evaluated=log.tickers_evaluated,
        orders_placed=log.orders_placed,
        orders_filled=log.orders_filled,
        skip_reason=log.skip_reason,
        error_message=log.error_message,
    )


# ──────────────────────────────────────────────
# Performance
# ──────────────────────────────────────────────

# ──────────────────────────────────────────────
# TradingDecision (Step 3: 승인 모드)
# ──────────────────────────────────────────────

class TradingDecisionResponse(BaseModel):
    id: UUID
    strategy_id: UUID
    account_id: UUID
    ticker: str
    action: str
    confidence: int
    reason: str
    suggested_quantity: Decimal | None
    executed: bool
    blocked_reason: str | None
    approval_status: str | None
    approval_expires_at: datetime | None
    created_at: datetime


def decision_to_response(decision: TradingDecisionModel) -> TradingDecisionResponse:
    return TradingDecisionResponse(
        id=decision.id,
        strategy_id=decision.strategy_id,
        account_id=decision.account_id,
        ticker=decision.ticker,
        action=decision.action,
        confidence=decision.confidence,
        reason=decision.reason,
        suggested_quantity=decrypt_decimal_optional(decision.suggested_quantity),
        executed=decision.executed,
        blocked_reason=decision.blocked_reason,
        approval_status=decision.approval_status,
        approval_expires_at=decision.approval_expires_at,
        created_at=decision.created_at,
    )


# ──────────────────────────────────────────────
# Performance
# ──────────────────────────────────────────────

class TradingPerformanceResponse(BaseModel):
    total_trades: int
    winning_trades: int
    losing_trades: int
    win_rate: Decimal
    total_realized_pnl: Decimal
    total_unrealized_pnl: Decimal
    initial_capital: Decimal
    current_value: Decimal
    return_rate: Decimal


# ──────────────────────────────────────────────
# AdaptiveRule (Step 4: 모듈 C)
# ──────────────────────────────────────────────

class AdaptiveRuleResponse(BaseModel):
    id: UUID
    strategy_id: UUID
    version: int
    rule_text: str
    rationale: str | None
    generated_from: str | None
    is_active: bool
    created_at: datetime
    expires_at: datetime | None


class AdaptiveRuleUpdate(BaseModel):
    is_active: bool


def adaptive_rule_to_response(rule: AdaptiveRuleModel) -> AdaptiveRuleResponse:
    return AdaptiveRuleResponse(
        id=rule.id,
        strategy_id=rule.strategy_id,
        version=rule.version,
        rule_text=rule.rule_text,
        rationale=rule.rationale,
        generated_from=rule.generated_from,
        is_active=rule.is_active,
        created_at=rule.created_at,
        expires_at=rule.expires_at,
    )
